### ISSUE TITLE
IREE codegen allocates excessive zero-filled buffer for MobileSSD models

### ISSUE THREAD

**Issue by hcindyl** [google/iree#6826]
While running IREE compiler on MobileSSD models (with TFLite format), the codegen requires a large amount of buffer (4.7MB in our test case) that's all filled in 0s. The intermediate operations could be fused and the buffer could be squashed.

Please see examples of mobilenet SSD models in 

In tf1 tensorflow pb format:
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md#coco-trained-models
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md#open-images-trained-models

In tflite format
https://coral.ai/models/object-detection/#trained-models. (notice there is a [TFLite_Detection_PostProcess](https://github.com/tensorflow/tensorflow/blob/1c419b231b622bd9e9685682545e9064f0fbb42a/tensorflow/lite/kernels/detection_postprocess.cc) op at the end of the models here and should be stripped before passing into the IREE codegen)

In tf2 savedmodel format
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md

**Comment by MaheshRavishankar** [comment]
Thanks Cindy! The immediate fix for these are to change the way the TOSA -> Linalg lowering is done. AFAICS, the TOSA -> Linalg path is currently lowering conv + bias to

```
%0 = linalg.generic %bias {
    indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d3)>,
                     affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>],
    iterator_types = ["parallel", parallel", "parallel", "parallel"]}
    ins(%bias : tensor<?xf32>) outs(%init : tensor<?x?x?x?xf32>) {
    ^bb0(%arg0 : f32, %arg1 : f32):
      linalg.yield %arg0 : f32
    } -> tensor<?x?x?x?xf32>
%1 = linalg.*conv* ... outs(%0 : tensor<?x?x?x?xf32>)

```

so its doing a broadcast of the bias into the full tensor which is then fed to the `outs` parameter of the conv operation. This is not exactly wrong, but is not what is currently handled correctly. What is handled better is

```
%cst = constant 0 : f32
%0 = linalg.fill(%cst, %init) : f32, tensor<?x?x?x?xf32> -> tensor<?x?x?x?xf32>
%1 = linalg.*conv* ... outs(%0 : tensor<?x?x?x?xf32>
%2 = linalg.generic {
    indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>,
                     affine_map<(d0, d1, d2, d3) -> (d3)>,
                     affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>],
    iterator_types = ["parallel", parallel", "parallel", "parallel"]}
    ins(%bias : tensor<?xf32>) outs(%init : tensor<?x?x?x?xf32>) {
    ^bb0(%arg0 : f32, %arg1 : f32):
      linalg.yield %arg0 : f32
    } -> tensor<?x?x?x?xf32>
   
```

That would be the first thing to fix.

As a side note, what is done today is still "correct". I have some local changes that can fuse what is currently seen by IREE through the TOSA path as well. I am in process of landing those (Needs https://reviews.llvm.org/D108565 and this commit in IREE https://github.com/MaheshRavishankar/iree/commit/5e3600641e194318b9a078a25e89d24fe375ea74). With this you get things fused, but it actually didnt help the memory consumption. It seems to me that buffer allocation already is reusing buffers in a way that makes this fusion unnecessary from a memory stand point (still better to do this since it is more efficient).

**Comment by benvanik** [comment]
Mahesh's fix is the best one; my workaround in #6815 is going to requires a few workarounds to work so I'm going to take the time to burn down some tech debt before it goes in.

If someone gets the model in tosa .mlir before me and could attach to the issue that'd be helpful for tracking down other issues too!

**Comment by julianwa** [comment]
Pinging this thread for @hcindyl. @MaheshRavishankar are you still waiting on response from @rsuderman?

**Comment by hcindyl** [comment]
For the original SSD model that we first saw the problem, the total memory usage is still 11MB for the float version (770kB). The quantized version memory usage dropped from 4.3MB to 2.8MB for the original 230kB tflite.

**Comment by benvanik** [comment]
oh, does rob's change fold the transposes? in my copy I still have those - some are like 4.7MB intermediates in the float model, so that's gonna skew the numbers a bit :P

**Comment by hcindyl** [comment]
@benvanik The `.tflite` in the tarballs of https://github.com/google/iree/issues/6826#issuecomment-905974356 (quant) and https://github.com/google/iree/issues/6826#issuecomment-905908692 (float) are still up-to-date. You need to regenerate the mlir with the latest `iree-import-tflite` though.

The 11MB usage is from the hal memory statistics I gather during the runtime (thanks for the super useful tool :100: ).

I will put the quant model to iree-samples/tflitehub so we can track it better.

**Comment by MaheshRavishankar** [comment]
Lei has a patch upstream for it. Should land soon too

**Comment by MaheshRavishankar** [comment]
yup, from way back when some hacky experiments gave the same result. I think the transpose change would help. I am on hook for reviewing it. Let me get to that.

**Comment by benvanik** [comment]
Awesome! The nice thing is that it'll help performance everywhere too - the memory savings are just a bonus :P

**Comment by hcindyl** [comment]
> @hcindyl: where'd you get the tflite quant memory usage number of 230KB? That seems physically impossible - the first conv in the model produces a 311KB tensor. There are also depthwise convs of 1x162x162x96xi8 (2.4MB) - and unless we do that in-place (is that possible?) that means we'll have a lower-bound usage of ~4.5MB. If tflite doesn't have that then either their reporting is wrong or they are doing something major that we need to replicate :P

The 230KB refers to the original TFLite binary size, not the intermediate activate size. I don't have the number of TFLite kernel internal memory usage. I agree for the public models shared above we could be close to the lower-bound, but for the lower depth-multiplier model that we are testing, the largest intermediate tensor is 1x162x162x16xi8, whereas the memory usage is 2.8MB, so somewhere in the pipe the memory could be saved.

Anything I can set to test memory usage with the stream branch for different models?

**Comment by hcindyl** [comment]
The 2.8MB usage is reported from `iree_hal_allocator_statistics_fprint` on exit, with the input .tflite size of 230K.

**Comment by hcindyl** [comment]
Correct, we (currently) don't have the tflite memory usage stat collected.

**Comment by benvanik** [comment]
Nice! That's really helpful! There's some future things we'll be able to do to reduce the transient memory usage more - like allowing for donation of input buffers would get us 307KB for free.

Looking at the file sizes, the public model (`ssd_mobilenet_v2_face_quant.tflite`) is 5.511MB .tflite and 5.651MB .vmfb. These compress to 3.123MB and 3.030MB respectively, meaning there's still a lot of low entropy stuff in there - but that's future work. A quick check shows 206KB being the ELF which compresses to 55KB. Makes sense given we are generating at least 2x as many kernels as we need and not deduplicating nearly any of them.

So, padding and improvements to deduping should let us close that 140KB gap - there's really only 5-6 unique dispatches but we are generating code for 118 :) There's also a decent chunk of low hanging fruit (constants that are pre-padded we could store unpadded and expand at init time, some zero padding, etc). Next year sometime it'd be interesting to play with constant compression (decoded on the fly at startup) as a storage-size vs. memory-size tradeoff.

**Comment by hcindyl** [comment]
FYI, `ssd_mobilenet_v2_face_quant.tflite` at snapshot 686 results in the peak memory usage of 5.93MB
```
[[ iree_hal_allocator_t memory statistics ]]
  HOST_LOCAL:            0B peak /            0B allocated /            0B freed /            0B live
DEVICE_LOCAL:      6218144B peak /      6218144B allocated /      6218144B freed /            0B live
```

**Comment by benvanik** [comment]
vs the 13.8MB of tflite? 🤞

**Comment by hcindyl** [comment]
From TFLite benchmark tool
```bash
Graph: [mobile_ssd_v2_quant.tflite]
#threads used for CPU inference: [1]
Loaded model mobile_ssd_v2_quant.tflite
The input model file size (MB): 5.51052
Initialized session in 1.357ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=1 curr=2302523

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=50 first=2293956 curr=2243167 min=2224464 max=2307193 avg=2.2736e+06 std=20999

Inference timings in us: Init: 1357, First inference: 2302523, Warmup (avg): 2.30252e+06, Inference (avg): 2.2736e+06
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Memory footprint delta from the start of the tool (MB): init=2.67188 overall=14.3906
Overall peak memory footprint (MB) via periodic monitoring: 20.4492
```

**Comment by benvanik** [comment]
wow, that's fantastic! and it only gets smaller from here :)

**Comment by MaheshRavishankar** [comment]
Is this still an issue?

**Comment by hcindyl** [comment]
We have seen some regression of memory usage since https://github.com/google/iree/issues/6826#issuecomment-987349128, but we are tracking it on https://github.com/google/iree/issues/7896. 

Maybe we can close this issue?

**Comment by MaheshRavishankar** [comment]
@hcindyl I am closing this bug. Please reopen if needed.
